pkg/flowaggregator/flowaggregator.go

```diff
@@ -15,25 +15,263 @@
+	}
+	if err != nil {
+		return fmt.Errorf("Got error when initializing IPFIX exporting process: %v", err)
+	}
+	fa.exportingProcess = ep
+	fa.templateID = fa.exportingProcess.NewTemplateID()
+	templateSet := ipfix.NewSet(ipfixentities.Template, fa.templateID, false)
+
+	bytesSent, err := fa.sendTemplateRecord(templateSet)
+	if err != nil {
+		fa.exportingProcess = nil
+		fa.templateID = 0
+		return fmt.Errorf("Send template message failed, err: %v", err)
 	}
+	klog.V(2).Infof("Initialized exporting process and sent %d bytes size of template record", bytesSent)
+	return nil
 }
 
 func (fa *flowAggregator) Run(stopCh <-chan struct{}) {
-	// Place holder for Run() function, logic will be added in next PR
+	var collectAddr net.Addr
+	var cp ipfix.IPFIXCollectingProcess
+	var err error
+	if fa.collectingProtocol == "tcp" {
+		collectAddr, _ = net.ResolveTCPAddr("tcp", "0.0.0.0:4739")
+		cp, err = ipfix.NewIPFIXCollectingProcess(collectAddr, 65535, 0)
+	} else {
+		collectAddr, _ = net.ResolveUDPAddr("udp", "0.0.0.0:4739")
+		cp, err = ipfix.NewIPFIXCollectingProcess(collectAddr, 1024, 0)
+	}
+	if err != nil {
+		klog.Errorf("Error when create collecting process: %v", err)
+		return
+	}
+	fa.collectingProcess = cp
+
+	ap, err := ipfix.NewIPFIXAggregationProcess(fa.collectingProcess.GetMsgChan(), 2, aggregationFields)
+	if err != nil {
+		klog.Errorf("Error when create aggreagation process: %v", err)
+		return
+	}
+	fa.aggregationProcess = ap
+
+	exportTicker := time.NewTicker(fa.exportInterval)
+	defer exportTicker.Stop()
+	go fa.collectingProcess.Start()
+	go fa.aggregationProcess.Start()
+	for {
+		select {
+		case <-stopCh:
+			fa.collectingProcess.Stop()
+			fa.aggregationProcess.Stop()
+			if fa.exportingProcess != nil {
+				fa.exportingProcess.CloseConnToCollector()
+			}
+			return
+		case <-exportTicker.C:
+			if fa.exportingProcess == nil {
+				err := fa.initExportingProcess()
+				if err != nil {
+					klog.Errorf("Error when initializing exporting process: %v", err)
+					// There could be other errors while initializing exporting process other than connecting to IPFIX collector,
+					// therefore closing the connection and resetting the process.
+					if fa.exportingProcess != nil {
+						fa.exportingProcess.CloseConnToCollector()
+						fa.exportingProcess = nil
+					}
+					continue
+				}
+			}
+			err := fa.aggregationProcess.ForAllRecordsDo(fa.sendFlowKeyRecords)
+			if err != nil {
+				klog.Errorf("Error when sending flow records: %v", err)
+				// If there is an error when sending flow records because of intermittent connectivity, we reset the connection
+				// to IPFIX collector and retry in the next export cycle to reinitialize the connection and send flow records.
+				fa.exportingProcess.CloseConnToCollector()
+				fa.exportingProcess = nil
+				continue
+			}
+			for _, key := range fa.sentFlowKeys {
+				fa.aggregationProcess.DeleteFlowKeyFromMap(*key)
+			}
+			fa.sentFlowKeys = make([]*ipfixintermediate.FlowKey, 0)
```

Comment:
you can also write `fa.sentFlowKeys = nil` - this will match what happens the first time you call `sendFlowKeyRecords`, since the slice is indeed `nil` at that point

Reply:
`fa.sentFlowKeys` has been removed, we choose to delete the flowkey from map as soon as we send it if be sent successfully. In the future, we will delete only inactive flows when intermediate process supports it in go-ipfix.